docs/commit-messages.md

```diff
@@ -0,0 +1,386 @@
+
+We use tags, because the repository is relatively large and most commits only
+apply to small parts of it. The tags make it easy to identify relevance of a
+commit at a glance while scanning through a large number of commits.
+
+The general guidelines prescribe capitalising the commit header. We do not
+capitalise the tag or after the tag, but we do capitalise the (rare) cases where
+there is no tag in the commit header.
+
+## Header and Body
+
+There is good general advice on [commit message writing][commit-messages]
+available on the internet and it is as relevant to proofs as it is to code.
+You should read it, it's not long and it's good advice.
+
+Repeating some high-level points:
+
+- Use imperative voice (e.g. `proofs: rename lemma foo to bar`)
+- The header should be a short summary, focussing on "what"
+- The body should explain what is going on in more detail (also in imperative
+  voice), but much more importantly *why* it is going on (is `bar` more
+  consistent than `foo`? Is the name `foo` needed for something else? Does `bar`
+  just better describe what is going on?).
+- You are trying to explain things to your future self or a future colleague
+  5-10 years from now. You can assume knowledge of the repository in general,
+  but you should not assume specific context that is obvious to you right now,
+  but that will not be known to a different person 5 years from now.
+
+## Tags
+
+- We use tags to indicate which part of the repository the commit applies to,
+  and if it is architecture-specific, then to which architecture it applies to.
+
+- We do no usually use branch tags, because git branches are ephemeral and we
+  are using rebase branches for most of our work. The one exception is the `rt`
+  branch, which has been alive for over half a decade. For this branch, we allow
+  merge commits (from branch `master` into `rt` only), and we want to be able to
+  reconstruct a rebase history from that at the end of the branch's life.
+
+  This means, we do use the tag `rt` for commits that only make sense on this
+  branch. If you could apply the commit to the master branch directly (e.g. you're
+  adding a lemma to a library), it should not get the tag. Otherwise it should.
+
+### Tag Examples
+
+The main tags we use are mostly the directory names of the main proof something
+is in, e.g. `refine`, `crefine`, `sys-init`, `camkes`. For some of these, there
+are abbreviations, mainly `aspec` for the abstract spec and `ainvs` for the
+abstract invariants.
+
+For large directories that have logical sub parts, we use slashes, e.g.
+`lib/monads`. Not so much because the change is in that directory, but because
+we want to see that it's a library change and applies to the monad part of the
+library.
+
+If the change applies to many proofs, for instance large lemma renaming commits,
+we use tags such as `proof` and `spec`.
+
+We combine tags with `+` if a change applies to multiple parts, e.g.
+`clib+crefine` or `lib+proof`.
```

Comment:
`clib` is a subdirectory of `lib`, so maybe we should either use `lib/clib` or make an exception for `clib`. `clib` is used quite a few times below

Reply:
This is an interesting point and maybe we should discuss that one a bit. `CLib` is both a session in the global session name space and a `clib` is a subdirectory of `lib`. We don't say `proof/refine` etc either, so maybe we should just get rid of the slashes altogether. I've used them in the past for things that I thought too obscure/small for people to know or find quickly. I'd be open to just not doing that.

Reply:
`Monads` is a case where I've been using slashes recently, for example `lib/monads/nondet`. `Monads` is also a global session, although a much smaller one, so I can live with or without the `lib` part of the tag. The suffix is useful here though, I often do want to disambiguate which monad I am working on and it's easier to just put it in the tag instead of trying to organically mention it in the header.

Reply:
The slashes are visually a bit confusing, because they represent both paths and `or`. I guess in tags people shouldn't be saying `or`.
I'm fine with an exception for clib, even though it's physically placed under lib.

For consideration: as per my comment on `word_lib`, now that monads are getting upstreamed it might make sense to also make an exception for `monads` in that case. There is a bunch of work upcoming for concurrency, and so we'll be seeing `lib/monads/nondet` a fair bit, which could become `monads/nondet`.

Reply:
I'd be happy with that.